docs/en/reference/testing.rst

```diff
@@ -0,0 +1,95 @@
+Testing Guidelines
+===================
+
+In order to ensure high quality, all components of the Doctrine DBAL library are extensively covered with tests.
+
+Having the code covered with tests and running all tests against each individual code change helps prevent
+breakages of the library logic when its code changes.
+
+Additionally, tests:
+
+1. Help maintainers understand what problem the given code change is trying to solve.
+2. Make sure that the problem being solved needs to be solved in the DBAL.
+3. Document the proper usage of the DBAL APIs.
+
+Requirements
+------------
+
+1. Each pull request that adds new logic or changes the existing logic must have tests.
+
+    (!) Exceptions to the rule:
+
+    - Modifications to the keyword lists under the `Doctrine\DBAL\Platforms\Keywords` namespace
+      don't have to be covered with tests.
+
+2. The test that covers certain logic must fail without the changes in the code it covers.
+
+Unit Tests
+----------
+
+Unit tests are meant to cover the logic of a given unit (e.g. a class or a method) including the logic of its interaction
+with other units. In this case, the other units could be mocked.
+
+Unit tests are mostly welcomed for testing the logic that the DBAL itself defines (e.g. logging).
+
+Integration Tests
+-----------------
+
+Integration (a.k.a. functional) tests are required when the behavior under test is dictated by the logic
+defined outside of the DBAL. It could be:
+
+1. The underlying database platform.
+2. The underlying database driver.
+3. SQL syntax and the standard as such.
+
+It is important to have integration tests for the cases above. Unlike unit tests, they make the external components
+the source of truth and help make sure that the logic implemented in the DBAL is correct even if the external components
+change (e.g. a new version of a database platform is supported).
+
+When are Integration Tests not required?
+----------------------------------------
+
+Some cases cannot be reproduced with the existing integration testing suite. It could be the scenarios that involve
+multiple concurrent database connections, transactions, locking, performance-related issues, etc.
+
+In such cases, it is still important that a pull request fixing the issues is accompanied with a free-form test that
+reproduces the issue being fixed.
```

Comment:
Ok, how about this then:

```suggestion
In such cases, it is still important that a pull request fixing the issues is accompanied with a free-form reproducer that demonstrates the issue being fixed.
```

? I think the word `test` should be avoided here.

Reply:
If "test" is misleading in this case, let's use "reproducer".